server/routes/content-security-policy-middleware.js

```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const crypto = require('crypto');
+const assert = require('assert');
+
+const config = require('../lib/config');
+const matrixServerUrl = config.get('matrixServerUrl');
+assert(matrixServerUrl);
+
+function contentSecurityPolicyMiddleware(req, res, next) {
+  const nonce = crypto.randomBytes(16).toString('hex');
+
+  // Based on https://web.dev/strict-csp/
+  const directives = [
+    // Default to fully-restrictive and only allow what's needed below
+    `default-src 'none';`,
+    // Only only <script> and <style> tags that have the nonce provided
+    //
+    // To ensure compatibility with very old browser versions (4+ years), we add
+    // 'unsafe-inline' as a fallback. All recent browsers will ignore 'unsafe-inline' if
+    // a CSP nonce or hash is present. (via
+    // https://web.dev/strict-csp/#step-4-add-fallbacks-to-support-safari-and-older-browsers)
+    `script-src 'nonce-${nonce}' 'strict-dynamic' https: 'unsafe-inline';`,
+    `style-src 'nonce-${nonce}' 'strict-dynamic' https: 'unsafe-inline';`,
+    // Hydrogen uses a bunch of inline styles
+    `style-src-attr 'unsafe-inline';`,
```

Comment:
`style-src-attr` doesn't really have [support in Firefox](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src-attr#browser_compatibility) yet (only supported in Nightly).

How can we craft a directive to only allow `<style nonce="...">` and inline styles? This seems hard since all recent browsers will ignore 'unsafe-inline' if a CSP nonce or hash is present.

Reply:
We could use `style-src self 'unsafe-inline';` but that's not *better* than a `nonce`. We might have to settle for that 🤔

Reply:
I think it's reasonable to settle for that, given constraints.